### PR TITLE
Update Python to 3.8.8

### PR DESCRIPTION
**Pull request by lucyzhang929**
This PR
- updates the Python fwlinks to the new Python 3.8.8 packages
- adds checks for ADS Python 3.6.6 users and prompts to upgrade

Note: We are removing the bundle version (0.0.1) from the default python path. Fixes https://github.com/microsoft/azuredatastudio/issues/14622

### CODE REVIEW

File: extensions/machine-learning/src/common/utils.ts [hunk 1]
```diff
@@ -65,10 +65,24 @@ export function getPythonInstallationLocation(rootFolder: string) {
...
+		rootFolder,
+		'0.0.1',
+		process.platform === constants.winPlatform ? 'python.exe' : 'bin/python3');
+	let newPythonPath = path.join(
+		rootFolder,
 		process.platform === constants.winPlatform ? 'python.exe' : 'bin/python3');
+
+	// Note: If Python exists in both paths (which can happen if the user chose not to remove Python 3.6 when upgrading),
+	// then we want to default to using the newer Python version.
+	if (!fs.existsSync(newPythonPath) && !fs.existsSync(oldPythonPath) || fs.existsSync(newPythonPath)) {
```
Comment (barbaravaldez) [inline extensions/machine-learning/src/common/utils.ts:80]: Why are you using the sync version of fs.exists?

**Review by barbaravaldez — APPROVED** [review]

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 2]
```diff
@@ -520,7 +564,7 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
 		}
 	}
 
-	public async getInstalledPipPackages(pythonExePath?: string): Promise<PythonPkgDetails[]> {
+	public async getInstalledPipPackages(pythonExePath?: string, user: boolean = false): Promise<PythonPkgDetails[]> {
```
Comment (corivera) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:567]: Could we rework the "user" argument here? Maybe something like "checkUserPackages"? The current name is a little vague at a glance.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 3]
```diff
@@ -165,7 +175,28 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
 
 		try {
 			let pythonExists = await utils.exists(this._pythonExecutable);
-			if (!pythonExists || forceInstall) {
+			let upgradePython = false;
+			// Warn users that some packages may need to be reinstalled after updating Python versions
+			if (pythonExists && !this._usingExistingPython && await this.getInstalledPythonVersion(this._pythonExecutable) !== constants.pythonVersion) {
```
Comment (corivera) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:180]: Are we targeting just Python 3.8, or do want them to be _at least_ on Python 3.8? 3.9 is the current version, so just doing an equality comparison here would prompt for a version downgrade if they're using that.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 4]
```diff
@@ -257,16 +287,6 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
 					.on('close', async () => {
 						//unpack python zip/tar file
 						outputChannel.appendLine(msgPythonUnpackPending);
-						let pythonSourcePath = path.join(installPath, constants.pythonBundleVersion);
-						if (await utils.exists(pythonSourcePath)) {
-							try {
-								// eslint-disable-next-line no-sync
-								fs.removeSync(pythonSourcePath);
```
Comment (corivera) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:264]: Were there any issues with overwriting an existing install in the exact same place? (like if you manually set the New Install path to the 0.0.1 folder)

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 5]
```diff
@@ -464,6 +488,12 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
 		}
 
 		let isPythonInstalled = JupyterServerInstallation.isPythonInstalled();
+
+		// If the latest version of ADS-Python is not installed, then prompt the user to upgrade
+		if (isPythonInstalled && !this._usingExistingPython && await this.getInstalledPythonVersion(this._pythonExecutable) !== constants.pythonVersion) {
```
Comment (corivera) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:493]: Same comment about version equality comparisons and Python 3.9.

**Review by corivera — APPROVED** [review]

File: extensions/machine-learning/src/common/utils.ts (hunk 1 above)
Comment (Charles-Gagnon) [reply extensions/machine-learning/src/common/utils.ts:80]: Product code should use promised versions

File: extensions/notebook/package.json [hunk 6]
```diff
@@ -38,6 +38,11 @@
           "default": false,
           "description": "%notebook.useExistingPython.description%"
         },
+        "notebook.dontPromptPythonUpdate": {
+          "type": "boolean",
+          "default": "",
```
Comment (Charles-Gagnon) [inline extensions/notebook/package.json:43]: Should have a default here (and it should be a boolean)

File: extensions/notebook/src/common/notebookUtils.ts [hunk 7]
```diff
@@ -19,17 +19,9 @@ export class NotebookUtils {
 
 	constructor() { }
 
-	public async newNotebook(connectionProfile?: azdata.IConnectionProfile): Promise<azdata.nb.NotebookEditor> {
+	public async newNotebook(options?: azdata.nb.NotebookShowOptions): Promise<azdata.nb.NotebookEditor> {
 		const title = this.findNextUntitledEditorName();
 		const untitledUri = vscode.Uri.parse(`untitled:${title}`);
-		const options: azdata.nb.NotebookShowOptions = connectionProfile ? {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/common/notebookUtils.ts:25]: Why are you removing this? That seems unrelated to your changes

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 8]
```diff
@@ -32,11 +32,19 @@ const msgInstallPkgStart = localize('msgInstallPkgStart', "Installing Notebook d
...
 const msgWaitingForInstall = localize('msgWaitingForInstall', "Another Python installation is currently in progress. Waiting for it to complete.");
+const msgPythonVersionUpdatePrompt = localize('msgPythonVersionUpdatePrompt', "Python 3.8.8 is now available in Azure Data Studio. The current Python version (3.6.6) will be out of support in December 2021. Would you like to update to Python 3.8.8 now?");
+const msgRemoveOldPythonPrompt = localize('msgRemoveOldPythonPrompt', "Would you like to remove the Python 3.6.6 installation?");
+const msgPythonVersionUpdateWarning = localize('msgPythonVersionUpdateWarning', "Python 3.8.8 will be installed and will replace Python 3.6.6. Some packages may no longer be compatible with the new version or may need to be reinstalled. A notebook will be created to help you reinstall all pip packages. Would you like to continue with the update now?");
+function msgRemovePython366Error(errorMessage: string, oldPythonPath: string): string { return localize('msgRemovePython366Error', "Encountered error while removing Python 3.6.6 installation: {0}. You can manually remove Python 3.6.6 by deleting this folder: {1}.", errorMessage, oldPythonPath); }
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:38]: Is the version always going to 3.6.6? The user might have updated it. Probably fine just to say `while removing Python from '<path to python>'`

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 9]
```diff
@@ -32,11 +32,19 @@ const msgInstallPkgStart = localize('msgInstallPkgStart', "Installing Notebook d
...
 const msgPythonRunningError = localize('msgPythonRunningError', "Cannot overwrite an existing Python installation while python is running. Please close any active notebooks before proceeding.");
 const msgWaitingForInstall = localize('msgWaitingForInstall', "Another Python installation is currently in progress. Waiting for it to complete.");
+const msgPythonVersionUpdatePrompt = localize('msgPythonVersionUpdatePrompt', "Python 3.8.8 is now available in Azure Data Studio. The current Python version (3.6.6) will be out of support in December 2021. Would you like to update to Python 3.8.8 now?");
+const msgRemoveOldPythonPrompt = localize('msgRemoveOldPythonPrompt', "Would you like to remove the Python 3.6.6 installation?");
+const msgPythonVersionUpdateWarning = localize('msgPythonVersionUpdateWarning', "Python 3.8.8 will be installed and will replace Python 3.6.6. Some packages may no longer be compatible with the new version or may need to be reinstalled. A notebook will be created to help you reinstall all pip packages. Would you like to continue with the update now?");
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:37]: Version should use the constant you defined so we don't have to update a bunch of places in the future.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts (hunk 3 above)
Comment (Charles-Gagnon) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:180]: Agreed, should do a semver compare here

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 10]
```diff
@@ -165,7 +175,28 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
...
 			let pythonExists = await utils.exists(this._pythonExecutable);
-			if (!pythonExists || forceInstall) {
+			let upgradePython = false;
+			// Warn users that some packages may need to be reinstalled after updating Python versions
+			if (pythonExists && !this._usingExistingPython && await this.getInstalledPythonVersion(this._pythonExecutable) !== constants.pythonVersion) {
+				upgradePython = await vscode.window.showInformationMessage(msgPythonVersionUpdateWarning, yes, no) === yes;
+			}
+
+			if (!pythonExists || forceInstall || upgradePython) {
+				if (upgradePython) {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:185]: This check doesn't seem correct - since it means that we'll essentially ignore the values of pythonExists/forceInstall

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 11]
```diff
@@ -165,7 +175,28 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
...
+					let userInstalledPipPackages: PythonPkgDetails[] = await this.getInstalledPipPackages(this._pythonExecutable, true);
+
+					if (await this.getInstalledPythonVersion(this._pythonExecutable) === '3.6.6') {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:188]: Why checking for this specific version?

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 12]
```diff
@@ -530,7 +574,12 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
 				return [];
 			}
 
-			let cmd = `"${pythonExePath ?? this.pythonExecutable}" -m pip list --format=json`;
+			let cmd = '';
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:577]: Could set this to the common part - `"${pythonExePath ?? this.pythonExecutable}" -m pip list --format=json` and then only append the --user if user is specified.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 13]
```diff
@@ -713,11 +761,25 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
...
+			'0.0.1',
+			process.platform === constants.winPlatform ? 'python.exe' : 'bin/python3');
+		let newPythonPath = path.join(
 			pythonInstallPath,
-			useExistingInstall ? '' : constants.pythonBundleVersion,
 			process.platform === constants.winPlatform ? 'python.exe' : 'bin/python3');
+
+		// Note: If Python exists in both paths (which can happen if the user chose not to remove Python 3.6 when upgrading),
+		// then we want to default to using the newer Python version.
+		if (!fs.existsSync(newPythonPath) && !fs.existsSync(oldPythonPath) || fs.existsSync(newPythonPath)) {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:777]: Don't use sync fs functions in product code if you can help it - use the promise versions

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts (hunk 3 above)
Comment (lucyzhang929) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:180]: Since I am also checking `usingExistingPython`, users who are using their own Python 3.9 would not be prompted. Although I can change the comparison to at least, just in case someone unchecked `usingExistingPython` setting even though they are using their own Python.

File: extensions/notebook/src/common/notebookUtils.ts (hunk 7 above)
Comment (lucyzhang929) [reply extensions/notebook/src/common/notebookUtils.ts:25]: I changed the parameter type from azdata.IConnectionProfile to azdata.nb.NotebookShowOptions so that I can pass in initialContent. This is for creating the notebook that helps users reinstall pip packages.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts (hunk 10 above)
Comment (lucyzhang929) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:185]: This is for steps that are only applicable to updating Python (ex. prompting to remove old installation). The rest of the installation process starts at line 200.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts (hunk 13 above)
Comment (lucyzhang929) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:777]: I noticed quite a few places in the jupyter installation code that uses fs.existsSync(). @corivera  do you have context around why we used sync? 
https://github.com/microsoft/azuredatastudio/blob/ed1db07f8cd4671587e534ba0d76c2081e154755/extensions/notebook/src/jupyter/jupyterServerInstallation.ts#L730
https://github.com/microsoft/azuredatastudio/blob/ed1db07f8cd4671587e534ba0d76c2081e154755/extensions/notebook/src/jupyter/jupyterServerInstallation.ts#L757

File: extensions/machine-learning/src/common/utils.ts [hunk 14]
```diff
@@ -65,10 +65,24 @@ export function getPythonInstallationLocation(rootFolder: string) {
 }
 
 export function getPythonExePath(rootFolder: string): string {
```
Comment (llali) [inline extensions/machine-learning/src/common/utils.ts:67]: >getPythonExePath [](http://example.com/codeflow?start=16&length=16)

looks like this function is not even used. feel free to remove

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts (hunk 4 above)
Comment (lucyzhang929) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:264]: If the new install path is set to the 0.0.1 folder then Python 3.6 installation gets overwritten. I'll add a check to make sure the new path and old path are not the same before prompting to remove, otherwise we'll be removing the new Python 3.8 installation.

Comment (lucyzhang929) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:264]: Actually, looking at the folder again, the Python 3.8 files are just added and not overwriting the existing install. So I'll have to check if the folder is the same as before and remove the 3.6 installation first.

**Review by llali — APPROVED** [review]
:shipit:

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts (hunk 4 above)
Comment (lucyzhang929) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:264]: @chlafreniere  @corivera  In this scenario (user manually changes the folder to the 0.0.1) and in the future when we do another Python update (and the folder path stays the same), are we okay with just overwriting the old installation and not prompting the user for removing the old Python version? 
If we want to continue allowing the user to keep the old Python version then we would need to add the bundle version back into the path.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 15]
```diff
@@ -165,7 +175,28 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
...
+			}
+
+			if (!pythonExists || forceInstall || upgradePython) {
+				if (upgradePython) {
+					let userInstalledPipPackages: PythonPkgDetails[] = await this.getInstalledPipPackages(this._pythonExecutable, true);
+
+					if (await this.getInstalledPythonVersion(this._pythonExecutable) === '3.6.6') {
+						this._promptToRemoveOldPythonInstallation = true;
+						this._oldPythonInstallationPath = this._pythonExecutable;
+						// remove '0.0.1' from python executable path
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:191]: Explain why you need to do this

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts (hunk 13 above)
Comment (Charles-Gagnon) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:775]: Probably just old code, we weren't great about trying to enforce that originally.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 16]
```diff
@@ -780,6 +840,37 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
...
+				},
+				language_info: {
+					name: 'python3'
+				}
+			},
+			nbformat: 4,
+			nbformat_minor: 5
+		};
+
+		await vscode.commands.executeCommand('_notebook.command.new', {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:869]: Instead of calling the command like this it would be clearer to just have the functions in the Utils class just be module-level functions instead of being part of a class. There's no reason for the class to exist and it just makes things harder to follow. 

(this should probably be done in a separate PR though to keep this one focused on the changes being made)

**Review by Charles-Gagnon — APPROVED** [review]
Overall the changes seem fine but I didn't spend too much time going over the specifics of the upgrade logic.

As discussed though I think it would be best if this waited until after the next release to go in to reduce churn in the repo.

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts [hunk 17]
```diff
@@ -438,7 +465,21 @@ export class JupyterServerInstallation implements IJupyterServerInstallation {
 
 						this._installCompletion.resolve();
 						this._installInProgress = false;
-						await vscode.commands.executeCommand('notebook.action.restartJupyterNotebookSessions');
+						if (this._upgradeInProcess) {
+							// Pass in false for restartJupyterServer parameter since the jupyter server has already been shutdown
+							// when removing the old Python version on Windows.
+							if (process.platform === constants.winPlatform) {
+								await vscode.commands.executeCommand('notebook.action.restartJupyterNotebookSessions', false);
```
Comment (barbaravaldez) [inline extensions/notebook/src/jupyter/jupyterServerInstallation.ts:472]: Why is the logic different for Windows?

**Review by barbaravaldez — APPROVED** [review]

File: extensions/notebook/src/common/constants.ts [hunk 18]
```diff
@@ -13,10 +13,10 @@ export const extensionOutputChannelName = 'Notebooks';
 export const notebookCommandNew = 'notebook.command.new';
 
 // JUPYTER CONFIG //////////////////////////////////////////////////////////
-export const pythonBundleVersion = '0.0.1';
-export const pythonVersion = '3.6.6';
+export const pythonVersion = '3.8.10';
```
[the remark this answers is not in the document]
Comment (lucyzhang929) [reply extensions/notebook/src/common/constants.ts:16]: Python 3.8.10 was recently released so I updated from 3.8.8 to the latest 3.8.10 (also updated the fwlinks to the new packages).

File: extensions/notebook/src/jupyter/jupyterServerInstallation.ts (hunk 17 above)
Comment (lucyzhang929) [reply extensions/notebook/src/jupyter/jupyterServerInstallation.ts:472]: Windows does not allow us to delete the Python files when it is in use. So we first have to shut down active notebook sessions before trying to remove the old python installation. On mac/linux we can delete the old python installation even if there are python notebook sessions running. Once the installation is complete, the open notebook sessions are restarted with the new python installation.

**Review by corivera — APPROVED** [review]

**Review by VasuBhog — APPROVED** [review]
